Improve the parsing of remote url details in the URL Details endpoint to REST API

getdave opened the pull request:
Builds on the original implementation from https://github.com/WordPress/gutenberg/pull/18042 to add a better mechanism for parsing the information from the remote URL's response body HTML.

This helps to address feedback such as https://github.com/WordPress/gutenberg/issues/27762 and also sets things up nicely for parsing additional detail from the remote URL (eg: favicon, Open Graph meta...etc).

This PR focuses solely on utilising `DOMDocument` and friends to improve the parsing mechanic.

## Automated Testing Instructions

`npm run test-php` should run the unit tests which reside in `phpunit/class-wp-rest-url-details-controller-test.php`.

## Manual Testing Instructions

* Check out this PR.
* Run `npm run wp-env start` to boot the local testing env.
* Open the Gutenberg local testing env at `http://localhost:8889/wp-admin/` and login.
* Create a new Post.
* Choose one of the following options:

### Option 1

* Open devtools and in the `console` enter:

```
wp.apiFetch( { path: '/__experimental/url-details/?url=https://wordpress.org' } ).then( data => {
    console.log( data );
} );
```

...also try some non-English websites:

```
wp.apiFetch( { path: '/__experimental/url-details/?url=http://www.baidu.com' } ).then( data => {
    console.log( data );
} );
```


You'll see the response returned there.

### Option 2
* Open devtools and in the `console` enter `wpApiSettings.nonce`. You should see a valid nonce returned as a string.
* Copy the nonce.
* In your browser goto: `http://localhost:8889/wp-json/__experimental/url-details/?url=https://wordpress.org&_wpnonce=%YOUR_NONCE_HERE%` - be sure to replace the nonce value with that copied from the previous step.
* You should see a valid REST API response containing the contents of the title tag from `wordpress.org` (ie: `"Blog Tool, Publishing Platform, and CMS \u2014 WordPress.org"`).


## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

getdave committed:
Use DOMDocument to parse title from remote website response

getdave committed:
Update tests to expect non-encoded data

getdave committed:
Extract method for building DOMXpath object from remote url response body

getdave committed:
Handle UTF-8 safely

getdave commented:
cc @beaulebens You might be interested in this as a follow up to https://github.com/WordPress/gutenberg/pull/18042#discussion_r542841004

TimothyBJacobs commented:
For future reference, `Core REST API Task` is for REST API issues that need to be fixed in WordPress Core.

getdave committed:
Strip HTML from title tag.

getdave committed:
Check and handle and test for parsing errors

getdave committed:
Ellaborate on how we’re testing for responses that cause errors in DOMDocument::loadHTML

getdave commented:
I'd like some feedback on how we're suppressing errors that can be generated by `DOMDocument::loadHTML`. For example if you include `<section>` tags in the response data then it will throw an error. We are suppressing the error which doesn't seem to cause any problems and we can still parse the data.

getdave commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -137,6 +143,39 @@ public function parse_url_details( $request ) {
 		return apply_filters( 'rest_prepare_url_details', $response, $url, $request, $remote_url_response );
```
Note to self. Pass the  `$xpath` value to the filter so folks can use it to query additional parts of response.

TimothyBJacobs commented:
I'm not sure I have enough experience with `DOMDocument` to give specific feedback here.

swissspidy commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -137,6 +143,39 @@ public function parse_url_details( $request ) {
+		// Ensure UTF-8 is respected by using 'mb_convert_encoding'.
+		$loaded_status = $doc->loadHTML( mb_convert_encoding( $remote_url_response, 'HTML-ENTITIES', 'UTF-8' ) );
+
+		libxml_use_internal_errors( $errors );
```
You might also want `libxml_clear_errors()` before this.

swissspidy commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -137,6 +143,39 @@ public function parse_url_details( $request ) {
+			return new WP_Error( 'rest_invalid_content', __( 'Unable to parse HTML response from remote URL.', 'gutenberg' ) );
+		}
+
+		$xpath = new DOMXpath( $doc );
+
+		return $xpath;
```
Would it be easier to justreturn the `$doc` here? So devs are not limited to using XPath.

getdave commented:
Picking this up again.

getdave committed:
Clear error buffer

Addresses https://github.com/WordPress/gutenberg/pull/28791#discussion_r576705486

getdave committed:
Separate DOMDoc and XPath into distinct function calls

hellofromtonya commented:
`DOMDocument` is a powerful HTML parser but presents **_significant problems_** for WordPress sites:

- Hosting: the following extensions are required: DOM, libxml, and iconv
Guarding could be added to protect against missing extensions. However, an alternative parsing mechanism would be needed  to parse these webpages.

- Libxml versions before 2.8.0 have a [known bug](http://xmlsoft.org/news.html): HTML parser error with `<noscript>` in the `<head>`
A transformation can be included to handle these instances.

- While `DOMDocument` autofixes for malformed HTML, it is not accurate with closing nested `div`s, when an inner child `div` is missing its closing tag.  Rather, it places the missing closing tag in the wrong place, changing the element relationships and structure.
The problem is the most problematic. Why? A missing inner closing div causes changes to the document's structure.  See the problem in action here https://3v4l.org/ijrfW

Story time:
During my time at my last company, we rolled out a solution using `DOMDocument`. We discovered over 30% of the webpages parsed through it were badly malformed and caused inaccurate DOM building and, worst yet, broken webpages after processing.


Not a problem if:
- If accuracy of inner div structures is not required and the Document is not converted back to HTML (i.e. via `DOMDocument::saveHTML` - or other methods)
- Only allowing well-formed webpages

However, if the point is to ensure proper HTML parsing for processing the elements, the accuracy of autofixing the HTML needs to be improved, especially for the missing inner closing div.

getdave commented:
@hellofromtonya Thanks for the detailed explanation. This isn't something I had considered and so it's extremely helpful to have this context. Much appreciated.

> However, if the point is to ensure proper HTML parsing for processing the elements, the accuracy of autofixing the HTML needs to be improved, especially for the missing inner closing div.

What I will say is that this is definitely being used for _progressive enhancement_ of the UI. It is _not_ critical functionality. All we are doing is _attempting_ to parse the remote website to gain some metadata to display to the user when they enter a valid link. If the parsing fails for any reason then it is absolutely fine and the worst the user will see will be the fallback UI for the link (which is what they see currently in the block editor).

I suppose there is an argument that if this endpoint exists then users _might_ try to use it for more _detailed_ parsing of a remote URL, but that is not its intent. Perhaps we could document as such or limit the response payload size to avoid folks using it as a scraper? This endpoint only runs in the admin if you have suitable permissions.

> Libxml versions before 2.8.0 have a known bug: HTML parser error with <noscript> in the <head>
A transformation can be included to handle these instances.

I can look into this.

> Hosting: the following extensions are required: DOM, libxml, and iconv
Guarding could be added to protect against missing extensions. However, an alternative parsing mechanism would be needed to parse these webpages.

I assume if we test for these extensions and bail if they don't cut the mustard then that's ok? Again, as this is progressive enhancement we can just not return any data and the block editor will provide fallback. The user won't really notice.

With the above context would you still say this approach is a no go?

hellofromtonya commented:
Hey @getdave, thanks for providing more context for its use. Not a "no go" yet. Extracting metadata such as the `<title>` is doable with `DOMDocument`. 

What type of metadata will be extracted? 

The PR is using xpath to find the document's `<title></title>` element. This particular element could be fetched using regex instead, which would be less code, faster code (parsing the HTML takes time especially for larger web pages), and without the server setup and encoding issues with `DOMDocument`.

Will more metadata be extracted from the HTML?

getdave commented:
> Will more metadata be extracted from the HTML?

Yes. Moreover, it is possible to use a filter hook on the endpoint and parse any data you want from the remote URL response.

> What type of metadata will be extracted?

The ultimate goal would be for the default data set to include:

* `<title>` contents
* site icon - eg: favicon...etc
* meta description

> The PR is using xpath to find the document's <title></title> element. This particular element could be fetched using regex instead, which would be less code, faster code (parsing the HTML takes time especially for larger web pages), and without the server setup and encoding issues with DOMDocument.

Ironically using regex is what the endpoint _currently_ does to get the `<title>`:

https://github.com/WordPress/gutenberg/blob/164bef81386e96330128c6218a66f29c9ed8e865/lib/class-wp-rest-url-details-controller.php#L210-L216

I wrote it and used regex and folks suggested using a more reliable mechanism if I wanted to extract more complex data which is why I raised this follow-up.

Ultimately would you advise that we ditch `DOMDocument` and just use regex for the parsing?

cc @swissspidy who has been using `DOMDocument` and may have suggested that approach.

getdave commented:
Ok we're going to take this in a new direction and use regex as the simplest way to parse out the markup we need. As the endpoint is extensible, folks can still choose to use more advanced utilities for their own purposes but we shouldn't ship this as part of core.

Let's also only grab the `<head>` portion of the DOM to avoiding having to parse a potentially massive string of HTML in regex.

swissspidy commented:
No strong opinion here as long as there is a reasonable (& documented) way for plugins to replace current parsing with DOMDocument if they want to.

getdave commented:
Just putting this here so I don't forget it:

`^(?=.*href="|\'(.*\.ico.*?)"|\')(?=.*rel="|\'(shortcut|icon)"|\').*$`

getdave commented:
Closing in favour of https://github.com/WordPress/gutenberg/pull/31763
